src/pallene/flow.lua

```diff
@@ -0,0 +1,309 @@
+    clear_set(dest)
+    for _,src_i in ipairs(src_indices) do
+        local src = state_list[src_i].finish
+        for v, _ in pairs(src) do
+            dest[v] = true
+        end
+    end
+end
+
+local function make_state_list(block_list, flow_info)
+    local state_list = {}
+    local order = flow_info.order._tag
+    for block_i, block in ipairs(block_list) do
+        local state = flow.FlowState.Build({},{},{},{})
+        state_list[block_i] = state
+        flow_info.init_start(state.start, block_i)
+        if order == "flow.Order.Forward" then
+            for cmd_i = 1, #block.cmds do
+                flow_info.process_cmd(state, block_i, cmd_i)
+            end
+        elseif order == "flow.Order.Backwards" then
+            for cmd_i = #block.cmds, 1, -1  do
+                flow_info.process_cmd(state, block_i, cmd_i)
+            end
+        else
+            tagged_union.error(order)
+        end
+    end
+    return state_list
+end
+
+function flow.flow_analysis(block_list, flow_info)
+
+    local state_list = make_state_list(block_list, flow_info)
+
+    local succ_list = ir.get_successor_list(block_list)
+    local pred_list = ir.get_predecessor_list(block_list)
+
+    local block_order
+    local merge_src_list
+    local dirty_propagation_list
+    local order = flow_info.order._tag
+    if order == "flow.Order.Forward"  then
+        block_order = ir.get_successor_depth_search_topological_sort(succ_list)
+        merge_src_list = pred_list
+        dirty_propagation_list = succ_list
+    elseif order == "flow.Order.Backwards" then
+        block_order = ir.get_predecessor_depth_search_topological_sort(pred_list)
+        merge_src_list = succ_list
+        dirty_propagation_list = pred_list
+    else
+        tagged_union.error(order)
+    end
+
+    local dirty_flag = {} -- { block_id -> bool } keeps track of modified blocks
+    for i = 1, #block_list do
+        dirty_flag[i] = true
+    end
+
+    local first_block_i = block_order[1]
+
+    local function update_block(block_i)
+        local state = state_list[block_i]
+
+        -- first block's starting set is supposed to be constant
+        if block_i ~= first_block_i then
+            local src_indices = merge_src_list[block_i]
+            merge_sets(state_list, src_indices, block_i)
+        end
+
+        local dirty_propagation = dirty_propagation_list[block_i]
+        local state_changed = apply_gen_kill_sets(state)
+        if state_changed then
+            for _, i in ipairs(dirty_propagation) do
+                dirty_flag[i] = true
+            end
+        end
+    end
+
+    repeat
+        local found_dirty_block = false
+        for _,block_i in ipairs(block_order) do
+            if dirty_flag[block_i] then
+                found_dirty_block = true
+                -- CAREFUL: we have to clean the dirty flag BEFORE updating the block or else we
+                -- will do the wrong thing for auto-referencing blocks
```

Comment:
What does "the wrong thing" mean?

Reply:
Means we'll always set the dirty flag to false, even if it is set to true during update_block, which might happen if the block  jumps to itself. I'll write a better comment.